### PR TITLE
refine to adapt nydus performance test

### PR DESCRIPTION
**Pull request by sctb512**
```shell
Usage:
run.sh -o OPERATION -s SOURCE_REGISTRY -t TARGET_REGISTRY [other options]
[-o operation]                  available options are [ push convert run all draw ]
[-i images]                     images list
[-p images path]                file path that contains images list (line by line)
[-s source registry]            source registry for pulling image
[-t target registry]            arget registry for pushing image
[-r round number]               number of round to run hellobench
[-d result directory]           directory to store raw result data
[-k skip finished test]         skip images that already finisned (in $RESULT_DIR/$RESULT_FILE)
```

- push image to local registry
```shell
./run.sh -o push -i alpine busybox -s registry/namespace -t target_registry/namespace
```

- convert image to nydus/stargz format and push to local registry
```shell
./run.sh -o convert -i alpine busybox -s registry/namespace -t target_registry/namespace
```

- run hellobench, `-r` indicates the round number.
```shell
./run.sh -o run -i alpine busybox -t target_registry/namespace
./run.sh -o run -i alpine busybox -t target_registry/namespace -r 1
```

Signed-off-by: Bin Tang <tangbin.bin@bytedance.com>

### CODE REVIEW

File: draw.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+import os
+import json
+import pandas as pd
+import shutil
+import matplotlib.pyplot as plt
+from pandas.api.types import CategoricalDtype
+import argparse
+
+parser = argparse.ArgumentParser(description='Handle data of hello bench to csv and png')
```
Comment (changweige) [inline draw.py:11]: Please use `black` as formatter  and `flake8` as static linter

**Comment by changweige** [comment]
It would be better if it can show the total time costed by pull + create + run
```
{"repo": "gcc", "bench": "gcc:nydusv6", "pull_time": 0.9336190223693848, "create_time": 0.8260679244995117, "run_time": 9.19385814666748}
```

**Comment by changweige** [comment]
Can we have a `hello.py` option to control cleaning up containers and images? So we have to a change to query nydusd running metrics to investigate performance issue

File: draw.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+                data.plot(kind="bar", x="type", rot=0,
+                          title="image: " + data_series["image"] + "  (" + data_series["type"] + ")", legend=False)
+                plt.xlabel(None)
+                plt.subplots_adjust(left=0.1, bottom=0.1, right=0.9, top=0.9)
+                plt.savefig(picture_path + "/" +
+                            data_series["image"].replace(":", "-") + "_" + data_series["type"] + ".png")
+
+
+def draw_all():
+    all_data_pd = pd.read_csv(result_dir + "/" + "all_mean.csv", index_col=0)
```
Comment (changweige) [inline draw.py:152]: Try use `path.join`

File: hello.py [hunk 3]
```diff
@@ -386,180 +399,460 @@ def image_ref(self, repo):
...
+            return optimize_cmd
+        elif repo in BenchRunner.CMD_STDIN:
+            optimize_cmd = self.optimize_cmd_stdin_cmd(repo, optimize_bin, accessed_files_dir, target_registry + "/" + target_repo, period)
+            print(optimize_cmd)
+            
+            # p = subprocess.Popen(optimize_cmd, shell=True, stdin=subprocess.PIPE,
+            # stdout=sys.stdout, stderr=sys.stdout, bufsize=0)
+            # print(BenchRunner.CMD_STDIN[repo].stdin)
+            # stdin=BenchRunner.CMD_STDIN[repo].stdin+'\nexit\n'
+            # p.communicate(stdin.encode())
```
Comment (changweige) [inline hello.py:838]: Please remove dead code if they not needed in the future

File: hello.py [hunk 4]
```diff
@@ -386,180 +399,460 @@ def image_ref(self, repo):
...
+        assert(rc == 0)
+        cmd = md = 'nerdctl --snapshotter %s rmi -f %s' % (self.snapshotter, image_ref)
+        print(cmd)
+        rc = os.system(cmd)
+        assert(rc == 0)
+    
+    def optimize_echo_hello_cmd(self, repo, optimize_bin, accessed_files_dir, target, period):
+        args = '--args=\'["%s"]\'' % "echo hello"
+        entry = '--entrypoint=\'["/bin/sh", "-c"]\''
+        cmd = ('%s record --period=%s --file=%s/%s.txt %s %s %s/%s:latest' %
```
Comment (changweige) [inline hello.py:750]: Try python 3 formatted string literals f-strings. It will be neater and easier

File: hello.py [hunk 5]
```diff
@@ -386,180 +399,460 @@ def image_ref(self, repo):
...
+        entry = '--entrypoint=\'["/bin/sh", "-c"]\''
+        args = ""
+        if BenchRunner.CMD_ARG_WAIT[repo].arg != "":
+            args = args + '--args=\'["%s"]\'' % BenchRunner.CMD_ARG_WAIT[repo].arg.replace('"', '\\\"').replace('\'', '\'"\'"\'')
+        if len(BenchRunner.CMD_ARG_WAIT[repo].env) > 0:
+            env = " ".join(["-e %s=%s" % (k, v) for k, v in BenchRunner.CMD_ARG_WAIT[repo].env.items()])
+            args = args + ' --env=\'["%s"]\'' % env.replace('"', '\\\"').replace('\'', '\'"\'"\'')
+        
+        # if BenchRunner.CMD_ARG_WAIT[repo].waitline != "":
+        #     optimize_bin = "%s --wait-on-line=\"%s\"" % (optimize_bin, BenchRunner.CMD_ARG_WAIT[repo].waitline)
```
Comment (changweige) [inline hello.py:789]: Try to remove dead code

File: hello.py [hunk 6]
```diff
@@ -386,180 +399,460 @@ def image_ref(self, repo):
...
+            args += ' --mount type=bind,src=%s,dst=%s,options=rbind' % (a,b)
+            
+        cmd = ('%s record --period=%s --file=%s/%s.txt %s %s %s/%s:latest' %
+            (optimize_bin, period, accessed_files_dir, repo , entry, args, target, repo))
+        return cmd
+
+    def optimize(self, bench, optimize_bin, target_registry, target_repo) -> str:
+        accessed_files_dir = "accessed_files"
+        if not os.path.exists(accessed_files_dir):
+            os.mkdir(accessed_files_dir)
```
Comment (changweige) [inline hello.py:816]: Try `os.makedirs(accessed_files_dir, exist_ok=True)`

File: hello.py [hunk 7]
```diff
@@ -386,180 +399,460 @@ def image_ref(self, repo):
...
+            # stdout=sys.stdout, stderr=sys.stdout, bufsize=0)
+            # print(BenchRunner.CMD_STDIN[repo].stdin)
+            # stdin=BenchRunner.CMD_STDIN[repo].stdin+'\nexit\n'
+            # p.communicate(stdin.encode())
+            
+            rc = os.system(optimize_cmd)
+            return optimize_cmd
+        elif repo in BenchRunner.CMD_ARG_WAIT:
+            optimize_cmd = self.optimize_cmd_arg_wait_cmd(repo, optimize_bin, accessed_files_dir, target_registry + "/" + target_repo, period)
+            print(optimize_cmd)
```
Comment (changweige) [inline hello.py:844]: Try `posix.path.join`

File: hello.py [hunk 8]
```diff
@@ -386,180 +399,460 @@ def image_ref(self, repo):
...
+            rc = os.system(optimize_cmd)
+            return optimize_cmd
+        elif repo in BenchRunner.CMD_URL_WAIT:
+            optimize_cmd = self.optimize_cmd_url_wait_cmd(repo, optimize_bin, accessed_files_dir, target_registry + "/" + target_repo, period)
+            print(optimize_cmd)
```
Comment (changweige) [inline hello.py:849]: Ditto

**Comment by sctb512** [comment]
> Can we have a `hello.py` option to control cleaning up containers and images? So we have to a change to query nydusd running metrics to investigate performance issue

Yes. Maybe we can store metrics for each test?

File: hello.py (hunk 6 above)
Comment (sctb512) [reply hello.py:816]: Get it . Thanks.

File: hello.py (hunk 5 above)
Comment (sctb512) [reply hello.py:789]: OK

File: hello.py [hunk 9]
```diff
@@ -661,7 +954,30 @@ def main():
...
+        type=str,
+        default="/usr/local/bin/optimize",
+    )
+    parser.add_argument(
+        "--target_registry",
+        type=str,
+        default="docker.io",
+    )
+    parser.add_argument(
+        "--target_repo",
```
Comment (changweige) [inline hello.py:975]: ditto

File: hello.py [hunk 10]
```diff
@@ -386,180 +399,460 @@ def image_ref(self, repo):
...
+        cmd = 'nerdctl --snapshotter %s rm -f %s' % (self.snapshotter, cid)
+        print(cmd)
+        rc = os.system(cmd)
+        assert(rc == 0)
+        cmd = md = 'nerdctl --snapshotter %s rmi -f %s' % (self.snapshotter, image_ref)
+        print(cmd)
+        rc = os.system(cmd)
+        assert(rc == 0)
+    
+    def optimize_echo_hello_cmd(self, repo, optimize_bin, accessed_files_dir, target, period):
```
Comment (changweige) [inline hello.py:747]: Can we move optimizer wrapper code to a new python file ?

File: hello.py [hunk 11]
```diff
@@ -386,180 +399,460 @@ def image_ref(self, repo):
...
+    def task_kill_cmd(self, cid):
+        cmd = 'nerdctl --snapshotter %s stop %s' % (self.snapshotter, cid)
+        return cmd
 
+    def cleanup(self, image_ref, cid):
+        print("Cleaning up environment for %s ..." % cid)
+        cmd = self.task_kill_cmd(cid)
+        print(cmd)
+        rc = os.system(cmd) # sometimes containers already exit. we ignore the failure.
+        cmd = 'nerdctl --snapshotter %s rm -f %s' % (self.snapshotter, cid)
```
Comment (changweige) [inline hello.py:738]: Can we use function `run` as `run(cmd, wait=True, verbose=False, shell=True)?

File: hello.py [hunk 12]
```diff
@@ -386,180 +399,460 @@ def image_ref(self, repo):
...
+        
+        print("Run time: %f s" % run_time)
+        
+        self.cleanup(image_ref, cid)
+        
+        return pull_time, create_time, run_time
+
+    def run_cmd_arg_wait(self, repo, runargs):
+        image_ref = self.image_ref(repo)
+        cid = repo.replace(":", "-")
```
Comment (changweige) [inline hello.py:492]: Is cid short for container id?

**Comment by changweige** [comment]
It's better not to commit binary file to repo

File: hello.py [hunk 13]
```diff
@@ -386,180 +399,460 @@ def image_ref(self, repo):
...
         image_ref = self.image_ref(repo)
-        docker = self.docker
-        docker.set_image(image_ref).run(run_cmd_args="echo hello")
+        cid = repo.replace(":", "-")
+        
+        pull_cmd = self.pull_cmd(image_ref)
+        print(pull_cmd)
+        
+        print("Pulling image %s ..." % image_ref)
+        start_pull = datetime.now()
```
Comment (changweige) [inline hello.py:408]: try to use python context manager to reduce tons of code 
```
def get_current_time():
    return datetime.datetime.now()


def delta_time(t_end, t_start):
    delta = t_end - t_start
    return delta.total_seconds(), delta.microseconds


@contextlib.contextmanager
def timer(slogan):
    start = get_current_time()
    try:
        yield
    finally:
        end = get_current_time()
        sec, usec = delta_time(end, start)
        logging.info("%s, Takes time %u.%u seconds", slogan, sec, usec // 1000)

```

File: hello.py (hunk 12 above)
Comment (sctb512) [reply hello.py:492]: Yes.

File: hello.py (hunk 10 above)
Comment (sctb512) [reply hello.py:747]: The arguments for optimization are from `BenchRunner`, which is a class in `hello.py`.

File: hello.py [hunk 14]
```diff
@@ -103,14 +105,40 @@ def tmp_copy(src):
...
+    return datetime.now()
+
+
+def delta_time(t_end, t_start):
+    delta = t_end - t_start
+    return delta.total_seconds(), delta.microseconds
+
+
+@contextmanager
+def timer(slogan):
```
Comment (changweige) [inline hello.py:118]: Shall we rename `slogan` to `cmd`?
slogan is a indicator telling what is being done in the context manager

File: hello.py [hunk 15]
```diff
@@ -103,14 +105,40 @@ def tmp_copy(src):
...
+    start = get_current_time()
+    try:
+        rc = os.system(slogan)
+        assert rc == 0
+        end = get_current_time()
+        sec, usec = delta_time(end, start)
+        yield sec + usec / 1e9
```
Comment (changweige) [inline hello.py:125]: 1e6?

Comment (sctb512) [reply hello.py]: Yes, fixed.

File: hello.py (hunk 14 above)
Comment (sctb512) [reply hello.py]: Done.

File: hello.py (hunk 7 above)
Comment (sctb512) [reply hello.py]: Already fixed it.

File: hello.py (hunk 4 above)
Comment (sctb512) [reply hello.py]: Done.

File: hello.py (hunk 3 above)
Comment (sctb512) [reply hello.py]: Done.

File: hello.py [hunk 16]
```diff
@@ -661,7 +954,30 @@ def main():
...
-
+    
+    parser.add_argument(
+        "--optimize",
+        type=bool,
+        choices=[True, False],
+        default=False,
+    )
+    parser.add_argument(
+        "--optimize_bin",
```
[the remark this answers is not in the document]
Comment (sctb512) [reply hello.py]: Fixed.

**Review by changweige — APPROVED** [review]

**Pull request merged**